src/de.rs

```diff
@@ -79,31 +85,47 @@ impl<'de> de::MapAccess<'de> for ObjectAccess {
     type Error = Error;
 
     fn next_key_seed<K: de::DeserializeSeed<'de>>(&mut self, seed: K) -> Result<Option<K::Value>> {
-        Ok(match self.fields.get(0) {
-            Some(&field) => Some(seed.deserialize(str_deserializer(field))?),
+        debug_assert!(self.next_value.is_none());
+
+        Ok(match self.fields.split_first() {
+            Some((field, fields)) => {
+                self.fields = fields;
+                let value = self.obj.get(static_str_to_js(field));
+                if value.is_undefined() {
+                    self.next_key_seed(seed)?
+                } else {
+                    self.next_value = Some(Deserializer::from(value));
+                    Some(seed.deserialize(str_deserializer(field))?)
+                }
+            }
             None => None,
         })
     }
 
     fn next_value_seed<V: de::DeserializeSeed<'de>>(&mut self, seed: V) -> Result<V::Value> {
-        let (field, fields) = self.fields.split_first().unwrap();
-        self.fields = fields;
-        let value = self.obj.get(static_str_to_js(field));
-        seed.deserialize(Deserializer::from(value))
+        let value = self.next_value.take().unwrap();
+        seed.deserialize(value)
     }
 
     fn next_entry_seed<K: de::DeserializeSeed<'de>, V: de::DeserializeSeed<'de>>(
         &mut self,
         kseed: K,
         vseed: V,
     ) -> Result<Option<(K::Value, V::Value)>> {
+        debug_assert!(self.next_value.is_none());
+
         Ok(match self.fields.split_first() {
             Some((&field, fields)) => {
                 self.fields = fields;
-                Some((
-                    kseed.deserialize(str_deserializer(field))?,
-                    vseed.deserialize(Deserializer::from(self.obj.get(static_str_to_js(field))))?,
-                ))
+                let value = self.obj.get(static_str_to_js(field));
+                if value.is_undefined() {
+                    self.next_entry_seed(kseed, vseed)?
```

Comment:
The only concern I have here is whether this will break some existing cases.

For example, what if value was a unit struct or, say, an `Option<something>`? Previously `undefined` would be passed all the way to the corresponding type's deserializer, but now I'm guessing it will error out due to missing field entry?

Reply:
To clarify - it might be still acceptable if there's truly no way around it, but I'd like to first understand if there's a way we can avoid breaking existing usecases.